Contrib/CDS Illumio

crestdatasystems opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Docker Image PR: https://github.com/demisto/dockerfiles/pull/9453

- Retrieves the traffic analysis of a particular port/protocol.
- Creates segmentation rule-sets for workloads.
- Create enforcement boundaries for micro-segmentation.
- These steps collectively provide the functionality of blocking a particular port.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [x] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

crestdatasystems commented:
Unit tests are failing because DockerImage https://github.com/demisto/dockerfiles/pull/9453 is not merged,

kgal-pan commented:
Unit test failing because of non-existant Docker image. @esharf - Please have a look at https://github.com/demisto/dockerfiles/pull/9453.

kgal-pan commented:
@crestdatasystems  - Please fill in contribution form. Your partner ID is `illumio-10397`. Also, please fix conflict in test conf.

crestdatasystems commented:
@esharf @kgal-pan I've updated contribs form and resolved the conflicts.

Can you please proceed further in the review as this is stale for 10 days?

esharf commented:
@crestdatasystems I'm in the middle of the review of this PR.
Hopefully, I will finish it today or tomorrow.
sorry for the inconvenience.

esharf commented on Packs/Illumio/Integrations/IllumioCore/IllumioCore.py:
```diff
@@ -0,0 +1,1232 @@
+
+        client = PolicyComputeEngine(url=base_url, port=port, org_id=org_id)
+        client.set_proxies(
+            http_proxy=proxy.get("http", None), https_proxy=proxy.get("https", None)
```
- if None is your default value then you don’t need to put it here (`{'a': 'b'}.get('c')` is equal to `{'a': 'b'}.get('c', None)`.
- more important is that in case the proxy is not settled, `handle_proxy()` will return `{'http': '', 'https': ''}` so you need to check it because it might lead to problems in case the SDK checks it explicitly against `None`.

esharf commented on Packs/Illumio/Integrations/IllumioCore/IllumioCore.py:
```diff
@@ -0,0 +1,1232 @@
+            else:
+                raise NotImplementedError("Command {} is not implemented".format(command))
+    except Exception as e:
+        demisto.error(traceback.format_exc())
```
```suggestion
```
`return_error` already takes care about it.

esharf commented on Packs/Illumio/Integrations/IllumioCore/IllumioCore.yml:
```diff
@@ -0,0 +1,1659 @@
+      isArray: false
+      name: protocol
+      predefined:
+      - TCP
+      - UDP
```
you are using upper case in a part of the time a lower in other.
consider to you a single standard here.

esharf commented on Packs/Illumio/Integrations/IllumioCore/IllumioCore.yml:
```diff
@@ -0,0 +1,1659 @@
+    - contextPath: Illumio.Rule.resolve_labels_as.consumers
+      description: Consumers resolve labels.
+      type: String
+  dockerimage: dev-illumio:1.1.2
```
please use the demisto docker image

kgal-pan replied:
```suggestion
  dockerimage: demisto/illumio:1.0.0.37679
```

esharf commented on Packs/Illumio/.pack-ignore:
```diff
@@ -0,0 +1,2 @@
+[file:IllumioCore_image.png]
+ignore=IM111
```
please try to fix it, it shouldn’t be something hard to fix. all you need is probably just reaching one of your designers. it should be very easy for them to fix it.

esharf commented on Packs/Illumio/Integrations/IllumioCore/IllumioCore.py:
```diff
@@ -0,0 +1,1232 @@
+                "illumio-workloads-list": workloads_list_command,
+                "illumio-enforcement-boundary-create": enforcement_boundary_create_command,
+                "illumio-enforcement-mode-update": update_enforcement_mode_command,
+                "illumio-ip-list-get": ip_list_get_command,
+                "illumio-ip-lists-get": ip_lists_get_command,
```
what is the difference between this 2 commands?

esharf commented on Packs/Illumio/Integrations/IllumioCore/IllumioCore.py:
```diff
@@ -0,0 +1,1232 @@
+    readable_output = prepare_workload_get_output(results)
+
+    return CommandResults(
+        outputs_prefix="Illumio.Workload",
```
```suggestion
        outputs_prefix="Illumio.Workloads",
```
take into account that the command might run multiple times and it might contain multiple results (yiu already have a path named `Illumio.Workloads"`
please fix it in the yml and readme files as well.

crestdatasystems replied:
Make sense, updated the same.

esharf commented on Packs/Illumio/Integrations/IllumioCore/IllumioCore.py:
```diff
@@ -0,0 +1,1232 @@
+    readable_output = prepare_ip_list_get_output(results)
+
+    return CommandResults(
+        outputs_prefix="Illumio.IPList",
```
```suggestion
        outputs_prefix="Illumio.IPLists",
```

esharf commented on Packs/Illumio/Integrations/IllumioCore/IllumioCore.py:
```diff
@@ -0,0 +1,1232 @@
+    readable_output = prepare_rule_create_output(response)  # type: ignore
+
+    return CommandResults(
+        outputs_prefix="Illumio.Rule",
```
```suggestion
        outputs_prefix="Illumio.RuleSet",
```
not sure about this one

crestdatasystems replied:
> not sure about this one

Rule and ruleset are different so won't be possible to update this change

esharf commented on Packs/Illumio/Integrations/IllumioCore/IllumioCore.py:
```diff
@@ -0,0 +1,1232 @@
+                "illumio-enforcement-mode-update": update_enforcement_mode_command,
+                "illumio-ip-list-get": ip_list_get_command,
+                "illumio-ip-lists-get": ip_lists_get_command,
+                "illumio-ruleset-create": ruleset_create_command,
+                "illumio-rule-create": rule_create_command,
```
what is the difference between this 2 commands?

crestdatasystems replied:
Ruleset is the superset where all the rules will be linked, where the rule is specific with more details. All rules should be linked to any ruleset.

esharf commented on Packs/Illumio/Integrations/IllumioCore/IllumioCore.py:
```diff
@@ -0,0 +1,1232 @@
+                "illumio-virtual-service-create": virtual_service_create_command,
+                "illumio-service-binding-create": service_binding_create_command,
+                "illumio-object-provision": object_provision_command,
+                "illumio-workload-get": workload_get_command,
+                "illumio-workloads-list": workloads_list_command,
```
what is the difference between this 2 commands?

crestdatasystems replied:
One will get the list of all workloads based on the arguments provided. Other will get the specific workload based on HREF provided.

esharf reviewed (changes requested):
some common comments:
- check for None before comparing optional variables with int variables (`None > 1` will raise a `TypeError`).
- please use `f'string: {s}` rather than `'string: {}'.format(s)`.
- try to use `dict_safe_get(data, ['a', 'b'])` rather than `data.get('a', {}).get('b')`.

esharf commented:
@idovandijk can you check the playbooks here?

idovandijk reviewed (commented):
@crestdatasystems thank you for the contribution.
I've dug into the documentation of the Illumio Core product to understand some of the terms that you are using. I'm basing my review on my understanding which may not be perfect, please correct me if I got something wrong.

### Rapid ransomware containment - Illumio
1. You have a task where you're getting all workloads, and then updating their enforcement mode to "selective". According to the documentation, only managed workloads can be modified to be in selective enforcement mode, but in your playbook you're not checking whether the workloads you found are managed before trying to change them to selective enforcement mode:
![image](https://user-images.githubusercontent.com/43602124/203544924-adf3a678-8955-4c4e-a71d-0b40ab69e96a.png)
2. If I understand correctly, if the traffic is allowed, then you are getting the workloads which were communicating in the said port and protocol. Then you are creating a virtual service which will serve as a label for the allowed service. Then you bind that service to the workloads where the traffic should be allowed. What I don't understand is the part where you're later creating a ruleset and a rule which are then applied to any IP. Can you clarify the objective of those tasks?

### Update enforcement mode - Illumio
3. Why is this playbook using global context + loops? Can it use inputs and outputs instead? Loops should be used when the subplaybook should make different decisions for each individual input, but it seems in this case that the playbook is linear and doesn't require looping. The downside of not using inputs is that this playbook is now hardcoded to update the enforcement mode to "selective" and won't be reusable for other use-cases.

After answering these points - can you suggest time windows for us to have a demo of the contribution? I'd like to understand better how Illumio works and the exact flow of the playbook.

Let me know if you have any questions.

crestdatasystems commented on Tests/conf.json:
```diff
@@ -121,7 +121,7 @@
         },
```
Done

crestdatasystems commented:
@esharf We've made all the changes suggested by you.

regarding the common comments, 
1. Client asked us specifically to not use the f-string. 
2. dict_safe_get is actually good idea that can be taken up in consideration for the upcoming integrations. However, RN changing data.get to dict_safe_get will result in a good amount of rework and may introduce regression so will skip this for this particular release.

crestdatasystems commented:
@esharf @idovandijk, We've observed that the unit test stage of CI trying to test the "test_module" function of the main file which results in the error which is expected because it is not the test case.

Can you please check and help why this has been changed previously it was not checking the "test_module" function of the integration script file.

https://app.circleci.com/pipelines/github/demisto/content/210043/workflows/afbaa700-a2f4-4de7-99df-d8357833b43b/jobs/603577/tests#failed-test-0

crestdatasystems commented:
@idovandijk For the playbook-related queries please find my answers.

1. I'll check for the code and API docs and will get back to you.

2. First we're getting traffic for any specific port then we're blocking all the traffic. Now if the traffic allowed is yes then it will create a rule and ruleset which will allow traffic for specific period of time. 

3. This playbook is part of the Main playbook(Rapid ransomware containment - Illumio). The purpose of this playbook is to loop certain tasks to update workloads enforcement mode. Also, we need certain exit conditions to stop the looping. To check that condition in the Main playbook, we are not able to access subplaybook's context that's why we have used global context.


Let me know if you have any other doubts or concerns.

esharf commented:
@crestdatasystems regarding the `test_nodule` problem in the unit tests.
this happened because pytest recognized this function as a test because it has a `test_` prefix please change it to something like `command_test_module`.
secondly please check my comments again there are a few comments you didn't address, please check them.
I’m available in [DFIR](https://dfircommunity.slack.com/team/UQWL4FQJ0) if you need any help with this.

crestdatasystems commented:
@esharf I've updated the image and addressed all the comments.

@esharf @idovandijk we're available to take a call tomorrow any time between 11:00 AM to 6 PM IST. Please let me know when we can have a call to complete this PR.

idovandijk commented:
Hi @crestdatasystems, I will be out of office for the next week. Since I don't want to have to delay this any further I will kindly ask that you let @efelmandar take the review from here. I will assign him as a reviewer.

crestdatasystems commented:
@esharf @efelmandar, let me know when we can schedule the demo for Illumio. We're available to have a demo call today or tomorrow anytime between 11:00 AM to 6 PM IST.

efelmandar commented:
@crestdatasystems Thank you again for the demo, I see that you finished fixing the playbooks according to the comments so we can go ahead with approving the PR, @esharf FYI
